conda/core/solve.py

```diff
@@ -1029,6 +1039,251 @@ def _prepare(self, prepared_specs):
+            always_update = [pkg.name for pkg in context.aggressive_update_packages
+                             if pkg.name in installed_names]
+            solver.add_jobs(always_update, api.SOLVER_UPDATE)
+        # 3b. Pin constrained packages in env
+        python_pin = self._pin_python(state)
+        if python_pin:
+            solver.add_pin(python_pin)
+        for pin in self._pinned_specs():
+            solver.add_pin(pin)
+
+        state["solver"] = solver
+        return solver
+
+    def _configure_solver_for_remove(self, state):
+        from mamba import mamba_api as api
+
+        solver_options = [
+            (api.SOLVER_FLAG_ALLOW_DOWNGRADE, 1),
+            (api.SOLVER_FLAG_ALLOW_UNINSTALL, 1)
+        ]
+        if context.channel_priority is ChannelPriority.STRICT:
+            solver_options.append((api.SOLVER_FLAG_STRICT_REPO_PRIORITY, 1))
+        solver = api.Solver(state["pool"], solver_options)
+
+        history = api.History(self.prefix)
+        history_map = history.get_requested_specs_map()
+        solver.add_jobs(
+            [s.conda_build_form() for s in history_map.values()],
+            api.SOLVER_USERINSTALLED,
+        )
+        specs = [s.conda_build_form() for s in self.specs_to_remove]
+        solver.add_jobs(specs, api.SOLVER_ERASE | api.SOLVER_CLEANDEPS)
+
+        state["solver"] = solver
+        return solver
+
+    def _collect_specs_to_add(self, state):
+        """
+        Reimplement the logic found in super()._add_specs(), but simplified.
+        """
+        # A. Explicitly requested packages via CLI (either update or install)
+        specs_to_add = list(self.specs_to_add)
+        # B. Implicitly requested via `update --all`: add everything in prefix
+        if self._command == 'update' and context.update_modifier == UpdateModifier.UPDATE_ALL:
+            for pkg in state["installed_pkgs"]:
+                if not pkg.name.startswith("__"):  # skip virtual packages
+                    specs_to_add.append(MatchSpec(pkg.name))
+        # C. Implicitly requested via `install|update --update-deps`
+        elif context.update_modifier == UpdateModifier.UPDATE_DEPS:
+            final_specs = specs_to_add
+            for spec in specs_to_add:
+                prec = next(p for p in state["installed_pkgs"] if p.name == spec.name)
+                for dep in prec.depends:
+                    ms = MatchSpec(dep)
+                    if ms.name != "python":
+                        final_specs.append(MatchSpec(ms.name))
+            specs_to_add = list(set(final_specs))
+        return [s.conda_build_form() for s in specs_to_add]
+
+    def _pin_python(self, state):
+        """
+        If Python is already present in the environment and the user didn't
+        include it in the requested packages, make sure it remains pinned to
+        the minor version.
+        """
+        installed_names = [rec.name for rec in state["installed_pkgs"]]
```

Comment:
`state["installed_pkgs"]` is used a few times, we should make this a cached property or something.

Reply:
Hm, `installed_pkgs` is only computed once while setting the state. After that it's only a dict lookup. What do you want to optimize?